# Исследование надежности заемщиков


Цель исследования — проверить надежность заемщиков по избранным характеристикам.

Многие факторы влияют на надежность заемщиков в глазах банка. Среди них - **количество детей у клиента, семейное положение, уровень дохода и цель кредита.**
Необходимо оценить зависимость возврата кредита в срок от вышеперечисленных характеристик клиентов.

Данные о клиентах находятся в файле data.csv. О качестве данных ничего не известно. Поэтому перед проверкой гипотез понадобится обзор данных.
На этапе предобработки необходимо исправить самые критичные ошибки данных - убрать пропуски, обработать дубликаты, аномальные значения.

Исследование пройдёт в три этапа:
1. Обзор данных.
2. Предобработка данных.
3. Проверка гипотез.

## Обзор данных

In [1]:
# импортируем библиотеку
import pandas as pd
#прочитаем файл
data = pd.read_csv('data.csv')

In [2]:
#откроем первые элементы таблицы
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

Видим, что в колонках `days_employed` и `total_income` есть пропуски.

В колонке `days_employed` есть аномальные значения - отрицательные величины.

В колонке `education` есть неявные дубликаты, которые могут мешать исследованию (в частности, группировке данных).


In [3]:
#посмотрим инфо о данных
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


Дополнительно подтверждаем наличие довольно значительного числа пропусков в `days_employed` и `total_income`. 

Также логично будет изменить тип данных в `days_employed` и `total_income` на целочисленный.

**Выводы**

Файл с данными анализа представлен таблицей с характеритиками о клиентах. В таблице указаны данные 21525 клиентов по 12 характеристикам (столбцам): children — количество детей в семье; days_employed — общий трудовой стаж в днях; dob_years — возраст клиента в годах; education — уровень образования клиента; education_id — идентификатор уровня образования, family_status — семейное положение; family_status_id — идентификатор семейного положения; gender — пол клиента; income_type — тип занятости; debt — имел ли задолженность по возврату кредитов; total_income — ежемесячный доход; purpose — цель получения кредита.

Представление данных имеет свои недостатки - аномалии, пропуски, дубликаты , которые уберем в ходе предобработки данных.
           

## Предобработка данных

### Обработка аномальных значений

Артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Заменим все отрицательные значения положительными с помощью метода `abs()`.

In [4]:
data['days_employed'] = data['days_employed'].abs()

В столбце `children` тоже есть аномальные значения. Уберем их.

In [5]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

In [6]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

Посмотрев медианные значения в колонке `total_income` по видам занятости, увидим, что медианные значения дохода некоторых групп клиентов банка  (безработных и пенсионеров) аномально большие. Исправить это сложно, поэтому оставим их как есть. Тем более, что мы не будем напрямую использовать эту колонку в исследовании.

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.137274
компаньон            1555.947387
пенсионер          365286.622650
предприниматель       520.848083
сотрудник            1572.328285
студент               578.751554
Name: days_employed, dtype: float64

### Обработка дубликатов

Обработаем неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведем их к нижнему регистру.

In [9]:
data['education'] = data['education'].str.lower()

Посмотрим, есть ли еще дубликаты в датасете.

In [10]:
data.duplicated().sum()

71

In [11]:
#есть , поэтому убираем строки-дубликаты
data = data.drop_duplicates()

### Удаление пропусков

In [12]:
#еще раз оценим количество пропусков 
data.isna().sum()

children               0
days_employed       2091
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2091
purpose                0
dtype: int64

Обработаем пропуски в столбце `total_income` и `days_employed`. Заменим пропуски на медианные значения по виду занятности `income_type`.    

In [13]:
# по доходу
for t in data['income_type'].unique():
    income_median = data.loc[(data['income_type'] == t), 'total_income'].median()
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = income_median

In [14]:
# по трудовому стажу
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

In [15]:
#проверим остались ли пропуски в столбце
data.isna().sum()
#заполнено

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

Заменим вещесттвенный тип данных в столбцах `days_employed` и `total_income` на целочисленный.

In [16]:
data['total_income'] = data['total_income'].astype(int)

In [17]:
data['days_employed'] = data['days_employed'].astype(int)

### Категоризация данных

На основании диапазонов, указанных ниже, создадим в датафрейме `data` столбец `total_income_category` с категориями:

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


In [18]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [19]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

Категоризируем цели кредита (значения в колонке `purpose`) по группам:

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследование данных

### 1. Есть ли зависимость между количеством детей и возвратом кредита в срок?

Чтобы это узнать, сгруппируем данные по количеству детей и посмотрим, много ли людей в разных группах имели задолженности по кредитам.

In [22]:
data.groupby('children')['debt'].sum()

children
0    1063
1     444
2     194
3      27
4       4
5       0
Name: debt, dtype: int64

Данная таблица показывает, что, чем больше детей у клиента, тем меньше вероятность задолженностей. Но эти результаты некорректны, так как, очевидно, что людей , например, с 1 ребенком больше, чем людей с 3-5 детьми. Поэтому необходимо сделать расчет относительно общего количества людей в каждой группе по количесву детей. 

In [23]:
#сколько людей в каждой группе
data['children'].value_counts()

0    14091
1     4808
2     2052
3      330
4       41
5        9
Name: children, dtype: int64

Количество людей в разных группах неодинаково и обратно пропорционально количеству детей.

In [24]:
#посчитали, сколько % из всех людей в каждой группе имели задолженности
debt_grouped_children = data.groupby('children')['debt'].sum()/data['children'].value_counts()*100 
debt_grouped_children

children
0    7.543822
1    9.234609
2    9.454191
3    8.181818
4    9.756098
5    0.000000
dtype: float64

In [25]:
#для удобства восприятия информации сделаем сводную таблицу
#сделаем это через функцию,тк будем использовать этот подход и для других колонок
# Создаем копию датасета, что бы не работать с исходным
temp = data.copy()
# Напишем функцию, так как все задачи аналогичны
def que(category):
    data_temp = temp.pivot_table(index=category, values='debt', aggfunc=['count', 'sum', 'mean'])
    data_temp.columns = ['Всего кредитополучателей', 'Всего должников', 'Доля должников']
    # Оформим таблицу цветным градиентом
    display(data_temp.style.format("{:.3f}").background_gradient(cmap='Blues', axis=0))
    
que('children')

**Вывод:** Зависимость между количеством детей и возвратом кредита в срок есть. Люди с 4-мя детьми чаще имеют задолженности относительно остальных групп. Меньше всех имеют задолженности люди без детей. Людей с пятью детьми не беру в сравнение, так как их всего 9 и, вероятно результат по этой группе нерепрезентативен.
В целом, люди без детей чаще возвращают кредит в срок.

### 2. Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [26]:
# оценим с помощью той самой сводной таблицы
que('family_status')

**Вывод:** зависимость между семейным положением и сроком возвращения кредита есть. Клиенты банка со статусом "Не женат / не замужем" чаще, чем остальные группы имеют задолженности. Вдовы и вдовцы, а также клиенты в разводе имеют меньше всего задолженностей. Оформленные семейные отношения или их прошлый опыт налагают определенную ответственность и приучают к финансовой дисциплине.

### 3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [27]:
#снова вызовем функцию. В этот раз выведем долю должников в группах с разным уровнем дохода
que('total_income_category')

Убираем из сравнения группы с маленькой выборкой - А, D и Е. 

*Вывод 1:*  
По двум выборкам с достаточным размером элементов можно сказать, что зависимость между доходом и погашением кредитов в срок есть. Так, люди с доходом С (50001–200000) чаще имеют задолженности по кредитам. Люди с доходом В (200001–1000000), чаще возвращают кредит вовремя. 


Из-за неравномерности количества элементов в категориях, получается мало данных для исследования. Попробуем применить метод `.qcut` для получения равномерных выборок.

In [28]:
#изменим категории на новые по методу .qcut
temp['total_income_category'] = pd.qcut(temp['total_income'], 5)

In [29]:
#посмотрим, что получается
que('total_income_category')

Если разделить общую выборку на 5 категорий методом `.qcut`, в каждой категории будет примерно одинаковое количество элементов. 

*Вывод 2:* Сравнивая новые более равномерные категории по задолженности, можно сделать вывод, что более надежными заемщиками являются люди с самым высоким доходом (около 200000 и более). Тем не менее люди с относительно низким доходом (примерно до 100 000 ) более надежны, чем клиенты с доходом от 100 000 до 200 000.


**Вывод общий по п.3.:** Закономерность между задолженностью и уровнем дохода есть: Наиболее надежными заемщиками являются люди с самым высоким доходом (около 200000 и более).

### 4. Как разные цели кредита влияют на его возврат в срок?

In [30]:
que('purpose_category')

**Вывод:** Зависимость между целью кредита и погашением в срок присутсвует. Чаще всего платят кредит в срок люди, взявшие его с целью операций с недвижимостью. Имеют задолженность чаще остальных клиенты, которые берут деньги на автомобиль и на образование.

## Общий вывод

1. Файл с данными анализа представлен таблицей с характеритиками о клиентах. В таблице указаны данные 21525 клиентов по следующим характеристикам (столбцам): `children` — количество детей в семье; `days_employed` — общий трудовой стаж в днях; `dob_years` — возраст клиента в годах; `education` — уровень образования клиента; `education_id` — идентификатор уровня образования, `family_status` — семейное положение; `family_status_id` — идентификатор семейного положения; `gender` — пол клиента; `income_type` — тип занятости; `debt` — имел ли задолженность по возврату кредитов; `total_income` — ежемесячный доход; `purpose` — цель получения кредита.


2. Перед началом анализа было необходимо избавиться от моментов, мешающих достоверному анализу. Аномальные значения в столбце `days_employed` (отрицательные значения трудового стажа) были заменены на соответсвующие положительные.  Пропуски в столбцах `days_employed` и `total_income`были заменены на медианные значения по типу занятости клиента. Тип данных `total_income` изменен на целочисленный. Неявные дубликаты в `education` обработаны за счет перевода в нижний регистр. Для данных в `total_income` и `purpose` была проведена категоризация и добавлены соответствующие столбцы в таблицуa `total_income_category` и `purpose_category` , чтобы эти данные были в удобном для анализа виде.


3. В ходе исследовательского анализа были проверены зависимости между возвратом кредита в срок и другими данными клиентов (количество детей, семейное положение, уровень дохода и цель кредита) и установлено следующее:

    a. Зависимость между количеством детей и возвратом кредита в срок есть. Люди с 4-мя детьми чаще имеют задолженности относительно остальных групп. Клиенты с 2-мя детьми реже вовремя отдают кредит нежели клиенты с 1-м и 3-мя детьми. Меньше всех имеют задолженности люди без детей. В целом, люди без детей чаще возвращают кредит в срок.

    b.Зависимость между семейным положением и сроком возвращения кредита есть. Клиенты банка со статусом "Не женат / не замужем" чаще, чем остальные группы имеют задолженности. Вдовы и вдовцы имеют меньше всего задолженностей.

    c. Закономерность между задолженностью и уровнем дохода есть:  наиболее надежными заемщиками являются люди с самым высоким доходом (около 200000 и более).

    d. Зависимость между целью кредита и погашением в срок есть. Чаще всего платят кредит в срок люди, взявшие его с целью операций с недвижимостью. Имеют задолженность чаще остальных клиенты, которые берут деньги на автомобиль или на образование.

Таким образом, наиболее надежным заемщиком является клиент без детей, был или находящийся в оформленных семейных отношениях, с доходом более 200 000 и который берет кредит для операций с недвижимостью.

Кроме того, во избежания пропусков в данных в дальнейшем, рекомендуется при онлайн заполнении заявления на получение кредита сделать проверку на не заполненые поля, которая не позвоит отправить заявку с пропусками в банк.